asset-registry/src/lib.rs

```diff
@@ -0,0 +1,243 @@
+	impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
+		fn build(&self) {}
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(T::WeightInfo::register_asset())]
+		#[transactional]
+		pub fn register_asset(
+			origin: OriginFor<T>,
+			metadata: AssetMetadata<T::Balance, T::CustomMetadata>,
+			asset_id: Option<T::AssetId>,
+		) -> DispatchResult {
+			let _ = T::AuthorityOrigin::ensure_origin(origin)?;
+
+			let (asset_id, metadata) = T::ProcessAsset::process_asset(asset_id, metadata)?;
+
+			Self::insert_metadata(&asset_id, &metadata)?;
+
+			Self::deposit_event(Event::<T>::RegisteredAsset { asset_id, metadata });
+
+			Ok(())
+		}
+
+		#[pallet::weight(T::WeightInfo::update_asset())]
+		#[transactional]
+		pub fn update_asset(
+			origin: OriginFor<T>,
+			metadata: AssetMetadata<T::Balance, T::CustomMetadata>,
+			asset_id: T::AssetId,
+		) -> DispatchResult {
+			let _ = T::AuthorityOrigin::ensure_origin(origin)?;
+
+			Self::update_metadata(&asset_id, &metadata)?;
+
+			Self::deposit_event(Event::<T>::UpdatedAsset { asset_id, metadata });
+
+			Ok(())
+		}
+
+		#[pallet::weight(T::WeightInfo::set_asset_location())]
+		#[transactional]
+		pub fn set_asset_location(
+			origin: OriginFor<T>,
+			asset_id: T::AssetId,
+			location: VersionedMultiLocation,
+		) -> DispatchResult {
+			let _ = T::AuthorityOrigin::ensure_origin(origin)?;
+
+			let old_metadata = Metadata::<T>::get(&asset_id).ok_or(Error::<T>::AssetNotFound)?;
+			let new_metadata = AssetMetadata {
+				location: Some(location.clone()),
+				..old_metadata
+			};
+			Self::update_metadata(&asset_id, &new_metadata)?;
+
+			Self::deposit_event(Event::<T>::SetLocation { asset_id, location });
+
+			Ok(())
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	fn update_metadata(
+		asset_id: &T::AssetId,
+		metadata: &AssetMetadata<T::Balance, T::CustomMetadata>,
+	) -> DispatchResult {
+		let old_metadata = Metadata::<T>::get(&asset_id).ok_or(Error::<T>::AssetNotFound)?;
+
+		if let Some(location) = old_metadata.location {
+			// remove the old location lookup
+			let location: MultiLocation = location.try_into().map_err(|()| Error::<T>::BadVersion)?;
+			MultiLocationLookup::<T>::remove(location);
+		}
+
+		Self::insert_metadata(asset_id, metadata)?;
+
+		Ok(())
+	}
+
+	pub fn fetch_metadata_by_location(
+		location: &MultiLocation,
+	) -> Option<AssetMetadata<T::Balance, T::CustomMetadata>> {
+		let asset_id = MultiLocationLookup::<T>::get(location)?;
+		Metadata::<T>::get(asset_id)
+	}
+
+	fn insert_metadata(
+		asset_id: &T::AssetId,
+		metadata: &AssetMetadata<T::Balance, T::CustomMetadata>,
+	) -> DispatchResult {
+		// if the metadata contains a location, set the MultiLocationLookup
+		if let Some(location) = metadata.location.clone() {
```

Comment:
I think this means when update asset metadata without changing location, the location is still going to be removed & inserted.

Not end of world as we shouldn't update asset metadata every block.

Reply:
You are right. However, since as you say, this shouldn't be called very frequently. We could optimize in the future but for now I prioritized code simplicity over performance